src/analyses/mayLocks.ml

```diff
@@ -1,18 +1,25 @@
-(** May lockset analysis (unused). *)
-
+(** May lockset analysis and analysis of double locking. *)
 open Analyses
+open GoblintCil
+module LF = LibraryFunctions
 
-module Arg =
+module Arg:LocksetAnalysis.MayArg =
 struct
-  module D = LockDomain.MayLockset
+  module D = LockDomain.MayLocksetNoRW
   module G = DefaultSpec.G
   module V = DefaultSpec.V
 
-  let add ctx l =
-    D.add l ctx.local
+  let add ctx (l,r) =
+    if D.mem l ctx.local then
+      match D.Addr.to_var_must l with
+      | Some v when ctx.ask (Queries.IsRecursiveMutex v)->
```

Comment:
`to_var_must` only accepts `NoOffset` to make this work only on direct `varinfo`s, but our locks are `lval`s, even in the non-symbolic case. Maybe the query could also work on lvalues, especially since the base analysis tracks mutex values arbitrarily deep in offsets.